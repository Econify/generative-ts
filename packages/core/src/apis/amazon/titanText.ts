import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/lib/Either.js";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { EjsTemplate } from "../../utils/ejsTemplate";

const templateSource = `{
  "inputText": "<%= prompt %>"
  <% if (typeof temperature !== 'undefined' || typeof topP !== 'undefined' || typeof maxTokenCount !== 'undefined' || typeof stopSequences !== 'undefined') { %>
  , "textGenerationConfig": {
    <% let comma = false; %>
    <% if (typeof temperature !== 'undefined') { %>
      "temperature": <%= temperature %>
      <% comma = true; %>
    <% } %>
    <% if (typeof topP !== 'undefined') { %>
      <% if (comma) { %>, <% } %> "topP": <%= topP %>
      <% comma = true; %>
    <% } %>
    <% if (typeof maxTokenCount !== 'undefined') { %>
      <% if (comma) { %>, <% } %> "maxTokenCount": <%= maxTokenCount %>
      <% comma = true; %>
    <% } %>
    <% if (typeof stopSequences !== 'undefined') { %>
      <% if (comma) { %>, <% } %> "stopSequences": <%- JSON.stringify(stopSequences) %>
    <% } %>
  }
  <% } %>
}`;

/**
 * @category Amazon Titan Text
 * @category Requests
 */
export interface AmazonTitanTextOptions extends ModelRequestOptions {
  temperature?: number; // float [0, 1.0]
  topP?: number; // float [1.00E-45, 1.0]
  maxTokenCount?: number; // int [0, 8000]
  stopSequences?: string[]; // must be one of: "|" | "User:"
}

/**
 * @category Amazon Titan Text
 * @category Templates
 */
export const AmazonTitanTextTemplate = new EjsTemplate<AmazonTitanTextOptions>(
  templateSource,
);

const AmazonTitanTextResponseCodec = t.type({
  inputTextTokenCount: t.number,
  results: t.array(
    t.type({
      outputText: t.string,
      completionReason: t.string,
      tokenCount: t.number,
    }),
  ),
});

/**
 * @category Amazon Titan Text
 * @category Responses
 */
export interface AmazonTitanTextResponse
  extends TypeOf<typeof AmazonTitanTextResponseCodec> {}

export function isAmazonTitanTextResponse(
  response: unknown,
): response is AmazonTitanTextResponse {
  return !isLeft(AmazonTitanTextResponseCodec.decode(response));
}

export interface AmazonTitanTextApi
  extends ModelApi<AmazonTitanTextOptions, AmazonTitanTextResponse> {}

/**
 *
 * ## Reference
 * [Amazon Titan Text](https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-titan-text.html)
 *
 * ## Providers using this API
 * - {@link createAwsBedrockModelProvider | AWS Bedrock}
 *
 * @category APIs
 * @category Amazon Titan Text
 * @category Provider: AWS Bedrock
 *
 */
export const AmazonTitanTextApi: AmazonTitanTextApi = {
  requestTemplate: AmazonTitanTextTemplate,
  responseGuard: isAmazonTitanTextResponse,
};
