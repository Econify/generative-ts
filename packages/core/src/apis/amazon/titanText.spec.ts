import { AmazonTitanTextApi, AmazonTitanTextOptions } from "./titanText";

function render(context: Omit<AmazonTitanTextOptions, "modelId">) {
  try {
    const rendered = AmazonTitanTextApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("AmazonTitanTextApi:", () => {
  /**
   * FewShotRequestOptions (prompt):
   */
  test("$prompt", () => {
    const rendered = render({
      $prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */
  test("all options", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      temperature: 0.7,
      topP: 0.9,
      maxTokenCount: 100,
      stopSequences: ["mock-stop-1", "mock-stop-2"],
    });

    expect(rendered).toMatchSnapshot();
  });
});
