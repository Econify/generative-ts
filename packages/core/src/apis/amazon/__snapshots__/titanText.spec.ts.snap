// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmazonTitanTextApi: all options 1`] = `
{
  "inputText": "mock-prompt",
  "textGenerationConfig": {
    "maxTokenCount": 100,
    "stopSequences": [
      "mock-stop-1",
      "mock-stop-2",
    ],
    "temperature": 0.7,
    "topP": 0.9,
  },
}
`;

exports[`AmazonTitanTextApi: prompt 1`] = `
{
  "inputText": "mock-prompt",
}
`;
